nomad/state/state_store.go

```diff
@@ -1486,6 +1491,36 @@ func (s *StateStore) nestedUpsertEval(txn *memdb.Txn, index uint64, eval *struct
 	return nil
 }
 
+// updateEvalModifyIndex is used to update the modify index of an evaluation that has been
+// through a scheduler pass. This is done as part of plan apply. It ensures that when a subsequent
+// scheduler workers process a re-queued evaluation it sees any partial updates from the plan apply.
+func (s *StateStore) updateEvalModifyIndex(txn *memdb.Txn, index uint64, evalID string) error {
+	// Lookup the evaluation
+	existing, err := txn.First("evals", "id", evalID)
+	if err != nil {
+		return fmt.Errorf("eval lookup failed: %v", err)
+	}
+	if existing == nil {
+		// return if there isn't an eval with this ID.
+		// In some cases (like snapshot restores), we process evals that are not already in the state store.
```

Comment:
Snapshot restores don't go through the normal upsert codepath. They have there own, direct insertion to avoid these problems.